File: Easy/Python/125. Valid Palindrome.py
""" Runtime: 43 ms, faster than 74.40%
    Memory Usage: 14.2 MB, less than 71.74% """

class Solution(object):
    def isPalindrome(self, s):
        l,r=0, len(s)-1
        while(l<=r):
            if not s[l].isalnum():
                l+=1
            elif not s[r].isalnum():
                r-=1
            elif s[l].lower()!=s[r].lower():
                return False
            else:
                l+=1
                r-=1
        return True

File: Easy/Java/125. Valid Palindrome.java
//Runtime: 4 ms, faster than 80.95%
//Memory Usage: 43.8 MB, less than 50.61%

class Solution {
    public boolean isPalindrome(String s) {
        int l=0;
        int r = s.length()-1;
        while(l<=r){
            char a = s.charAt(l);
            char b = s.charAt(r);
            if (!Character.isAlphabetic(a) && !Character.isDigit(a)){
                l++;
            } else if (!Character.isAlphabetic(b) && !Character.isDigit(b)){
                r--;
            }
            else if (Character.compare(Character.toLowerCase(a),Character.toLowerCase(b))!=0){

                return false;
            }else{
              l++;
              r--;   
            }
        }
        return true;
    }
}